File: scripts/refactor_terms.js
const fs = require('fs');
const path = require('path');

const TARGET_DIR = "c:\\Users\\Keisha\\Documents\\caltrans\\new caltrans";

const BANNED_TERMS = [
    { pattern: /\bVendor\b/g, replacement: 'Small Business' },
    { pattern: /\bvendor\b/g, replacement: 'small business' },
    { pattern: /\bVENDOR\b/g, replacement: 'SMALL BUSINESS' },
    { pattern: /\bVendors\b/g, replacement: 'Small Businesses' },
    { pattern: /\bvendors\b/g, replacement: 'small businesses' },
    { pattern: /\bVENDORS\b/g, replacement: 'SMALL BUSINESSES' },
    
    { pattern: /\bSBE\b/g, replacement: 'Small Business' },
    { pattern: /\bsbe\b/g, replacement: 'small business' },
    { pattern: /\bSBEs\b/g, replacement: 'Small Businesses' },
    { pattern: /\bsbes\b/g, replacement: 'small businesses' },

    { pattern: /\bDBE\b/g, replacement: 'Small Business' },
    { pattern: /\bdbe\b/g, replacement: 'small business' },
    { pattern: /\bDBEs\b/g, replacement: 'Small Businesses' },
    { pattern: /\bdbes\b/g, replacement: 'small businesses' },

    { pattern: /\bDisadvantaged Business\b/g, replacement: 'Small Business' },
    { pattern: /\bdisadvantaged business\b/g, replacement: 'small business' },
    { pattern: /\bDisadvantaged business\b/g, replacement: 'Small business' },
    { pattern: /\bDisadvantaged Businesses\b/g, replacement: 'Small Businesses' },
    { pattern: /\bdisadvantaged businesses\b/g, replacement: 'small businesses' },
    { pattern: /\bDisadvantaged businesses\b/g, replacement: 'Small businesses' },
    
    { pattern: /\bAgency\b/g, replacement: 'Prime Contractor' },
    { pattern: /\bagency\b/g, replacement: 'prime contractor' },
    { pattern: /\bAGENCY\b/g, replacement: 'PRIME CONTRACTOR' },
    { pattern: /\bAgencies\b/g, replacement: 'Prime Contractors' },
    { pattern: /\bagencies\b/g, replacement: 'prime contractors' },
    { pattern: /\bAGENCIES\b/g, replacement: 'PRIME CONTRACTORS' }
];

const report = {
    files_modified: [],
    skipped_instances: [],
    ambiguous_cases: [],
    snippets: []
};

// Negative lookbehinds/lookaheads (simulated due to JS regex engine limits)
function isSafeContext(text, matchIndex, matchLength) {
    const start = matchIndex;
    const end = matchIndex + matchLength;
    
    // Check if inside a URL or property like variable-name, variable_name
    if (start > 0 && ['-', '_', '.', '/'].includes(text[start-1])) return false;
    if (end < text.length && ['-', '_', '.', '/'].includes(text[end])) return false;
    
    // Check if preceding or succeeding characters are letters/digits (camelCase)
    if (start > 0 && /[a-zA-Z]/.test(text[start-1])) return false;
    if (end < text.length && /[a-zA-Z]/.test(text[end])) return false;
    
    return true;
}

function processFile(filepath) {
    try {
        let content = fs.readFileSync(filepath, 'utf-8');
        let originalContent = content;
        let modified = false;
        
        // We will do replacements manually to check safe context and keep track
        BANNED_TERMS.forEach(({ pattern, replacement }) => {
            let match;
            // Reset regex to start
            pattern.lastIndex = 0;
            // We use matchAll approach loop
            // We must collect matches first, process in reverse
            const matches = [];
            while ((match = pattern.exec(content)) !== null) {
                matches.push({
                    index: match.index,
                    str: match[0],
                    length: match[0].length
                });
            }
            
            for (let i = matches.length - 1; i >= 0; i--) {
                const { index, str, length } = matches[i];
                const snippetStart = Math.max(0, index - 30);
                const snippetEnd = Math.min(content.length, index + length + 30);
                const snippet = content.substring(snippetStart, snippetEnd).replace(/\n/g, ' ');
                
                if (isSafeContext(content, index, length)) {
                    content = content.substring(0, index) + replacement + content.substring(index + length);
                    modified = true;
                    if (report.snippets.length < 50) {
                        report.snippets.push(`✅ Replaced '${str}' -> '${replacement}' in ${path.basename(filepath)}: ...${snippet}...`);
                    }
                } else {
                    report.skipped_instances.push(`⏭️ Skipped '${str}' in ${path.basename(filepath)} due to unsafe context: ...${snippet}...`);
                }
            }
        });
        
        if (modified) {
            fs.writeFileSync(filepath, content, 'utf-8');
            report.files_modified.push(filepath);
        }
    } catch (e) {
        console.error(`Error processing ${filepath}: ${e.message}`);
    }
}

function walkDir(dir) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const fullPath = path.join(dir, file);
        if (['node_modules', '.git', '.github'].includes(file)) continue;
        
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
            walkDir(fullPath);
        } else if (['.html', '.js', '.json', '.sql', '.md', '.css'].includes(path.extname(fullPath))) {
            processFile(fullPath);
        }
    }
}

walkDir(TARGET_DIR);

const reportPath = path.join(TARGET_DIR, "TERMINOLOGY_REPORT.md");
let reportContent = "# Terminology Refactoring Report\n\n";
reportContent += "## Summary\n";
reportContent += `- Files Modified: ${report.files_modified.length}\n`;
reportContent += `- Instances Skipped: ${report.skipped_instances.length}\n\n`;

reportContent += "## Modified Files\n";
report.files_modified.forEach(file => {
    reportContent += `- ${file.replace(process.cwd(), '')}\n`;
});

reportContent += "\n## Diff Snippets (Sample)\n";
report.snippets.slice(0, 50).forEach(snippet => {
    reportContent += `- ${snippet}\n`;
});

reportContent += "\n## Skipped Instances (Code/URL variables)\n";
report.skipped_instances.slice(0, 50).forEach(skip => {
    reportContent += `- ${skip}\n`;
});

fs.writeFileSync(reportPath, reportContent, 'utf-8');
console.log("Refactoring complete. Report generated at TERMINOLOGY_REPORT.md");

File: scripts/refactor_terms.py
import os
import re
import json

TARGET_DIR = r"c:\Users\Keisha\Documents\caltrans\new caltrans"

BANNED_TERMS = {
    # Vendor
    re.compile(r'\bVendor\b'): 'Small Business',
    re.compile(r'\bvendor\b'): 'small business',
    re.compile(r'\bVENDOR\b'): 'SMALL BUSINESS',
    re.compile(r'\bVendors\b'): 'Small Businesses',
    re.compile(r'\bvendors\b'): 'small businesses',
    re.compile(r'\bVENDORS\b'): 'SMALL BUSINESSES',
    
    # SBE
    re.compile(r'\bSBE\b'): 'Small Business',
    re.compile(r'\bsbe\b'): 'small business',
    re.compile(r'\bSBEs\b'): 'Small Businesses',
    re.compile(r'\bsbes\b'): 'small businesses',
    
    # DBE
    re.compile(r'\bDBE\b'): 'Small Business',
    re.compile(r'\bdbe\b'): 'small business',
    re.compile(r'\bDBEs\b'): 'Small Businesses',
    re.compile(r'\bdbes\b'): 'small businesses',
    
    # Disadvantaged Business
    re.compile(r'\bDisadvantaged Business\b'): 'Small Business',
    re.compile(r'\bdisadvantaged business\b'): 'small business',
    re.compile(r'\bDisadvantaged business\b'): 'Small business',
    re.compile(r'\bDisadvantaged Businesses\b'): 'Small Businesses',
    re.compile(r'\bdisadvantaged businesses\b'): 'small businesses',
    re.compile(r'\bDisadvantaged businesses\b'): 'Small businesses',
    
    # Agency
    re.compile(r'\bAgency\b'): 'Prime Contractor',
    re.compile(r'\bagency\b'): 'prime contractor',
    re.compile(r'\bAGENCY\b'): 'PRIME CONTRACTOR',
    re.compile(r'\bAgencies\b'): 'Prime Contractors',
    re.compile(r'\bagencies\b'): 'prime contractors',
    re.compile(r'\bAGENCIES\b'): 'PRIME CONTRACTORS'
}

# Negative lookbehinds/lookaheads to prevent matching inside URLs, paths, or code variable syntax.
# We don't want to replace "vendorId" or "agency_name" or "vendor-profile.html"
def is_safe_context(text, start, end):
    # Check if inside a URL or property like variable-name, variable_name
    if start > 0 and text[start-1] in '-_./': return False
    if end < len(text) and text[end] in '-_./': return False
    
    # Check if preceding or succeeding characters are letters/digits (camelCase)
    if start > 0 and text[start-1].isalpha(): return False
    if end < len(text) and text[end].isalpha(): return False
    
    return True

report = {
    'files_modified': [],
    'skipped_instances': [],
    'ambiguous_cases': [],
    'snippets': []
}

def process_file(filepath):
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            original_content = f.read()
    except Exception as e:
        return
    
    content = original_content
    modified = False
    
    for pattern, replacement in BANNED_TERMS.items():
        matches = list(pattern.finditer(content))
        # Process in reverse so replacements don't mess up indices
        for match in reversed(matches):
            start, end = match.span()
            matched_str = match.group()
            
            # Context window for snippets/checks
            snippet_start = max(0, start - 30)
            snippet_end = min(len(content), end + 30)
            snippet = content[snippet_start:snippet_end].replace('\n', ' ')
            
            if is_safe_context(content, start, end):
                # Apply replacement
                content = content[:start] + replacement + content[end:]
                modified = True
                if len(report['snippets']) < 50: # Limit snippet collection
                    report['snippets'].append(f"✅ Replaced '{matched_str}' -> '{replacement}' in {os.path.basename(filepath)}: ...{snippet}...")
            else:
                report['skipped_instances'].append(f"⏭️ Skipped '{matched_str}' in {os.path.basename(filepath)} due to unsafe context: ...{snippet}...")
                
    if modified:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        report['files_modified'].append(filepath)

for root, dirs, files in os.walk(TARGET_DIR):
    if any(ignore in root for ignore in ['node_modules', '.git', '.github']):
        continue
    for file in files:
        if file.endswith(('.html', '.js', '.json', '.sql', '.md', '.css')):
            process_file(os.path.join(root, file))

# Write report
report_path = os.path.join(TARGET_DIR, "TERMINOLOGY_REPORT.md")
with open(report_path, 'w', encoding='utf-8') as f:
    f.write("# Terminology Refactoring Report\n\n")
    f.write("## Summary\n")
    f.write(f"- Files Modified: {len(report['files_modified'])}\n")
    f.write(f"- Instances Skipped: {len(report['skipped_instances'])}\n\n")
    
    f.write("## Modified Files\n")
    for file in report['files_modified']:
        f.write(f"- {file.replace(TARGET_DIR, '')}\n")
    
    f.write("\n## Diff Snippets (Sample)\n")
    for snippet in report['snippets'][:20]:
        f.write(f"- {snippet}\n")
        
    f.write("\n## Skipped Instances (Code/URL variables)\n")
    for skip in report['skipped_instances'][:50]:
        f.write(f"- {skip}\n")

print("Refactoring complete. Report generated at TERMINOLOGY_REPORT.md")
